### PR TITLE
[DTensor] min/max dim reduction fix

### LINKED ISSUE #168940: [DTensor] aten.max.dim returns wrong indices when using DTensor
### 🐛 Describe the bug

I found that current strategy of `aten.max.dim` may get incorrect indices output if sharded the dim for maximization.

Sample code:
```python
import torch
from torch.distributed.tensor import distribute_tensor, Shard

from torch.testing._internal.common_utils import run_tests
from torch.testing._internal.distributed._tensor.common_dtensor import DTensorTestBase, with_comms


class TestRegisterSharding(DTensorTestBase):
    @with_comms
    def test_max_dim(self):
        mesh = self.build_device_mesh()

        x = torch.randn(4, 4, device="cuda")

        max_value, max_indices = torch.max(x, dim=1)

        dist_x = distribute_tensor(x, mesh, [Shard(1)])

        dist_max_value, dist_max_indices = torch.max(dist_x, dim=1)

        print("x:", x)
        print("max_value:", max_value)
        print("max_indices:", max_indices)
        print("dist_max_value:", dist_max_value.full_tensor())
        print("dist_max_indices:", dist_max_indices.full_tensor())


if __name__ == "__main__":
    run_tests()
```

Result:
```python
x: tensor([[-1.6165,  0.5685, -0.5102, -0.9113],
        [-1.1555, -0.2262, -1.2891,  1.0654],
        [-0.7167, -0.5333,  0.2078, -0.9798],
        [ 0.7447, -0.2395,  0.2737,  0.0920]], device='cuda:0')
max_value: tensor([0.5685, 1.0654, 0.2078, 0.7447], device='cuda:0')
max_indices: tensor([1, 3, 2, 0], device='cuda:0')
dist_max_value: tensor([0.5685, 1.0654, 0.2078, 0.7447], device='cuda:0')
dist_max_indices: tensor([0, 0, 0, 0], device='cuda:0')
```

Each rank gets a shape(4, 1) local tensor to call `max.dim` in this case, and the local result of max indices is [0, 0, 0, 0]. The framework doesn't process the offset of index, which leads to an incorrect global result when the relavant dim is sharded.

Is there a good way to implement a strategy that supports sharding the index dim?

### Versions

torch v2.9.0

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @tianyu-l @XilunWu @SherlockNoMad

### PR DESCRIPTION
**Pull request by skpark-rh**
Fixes #168940

## Issue
The bug happens because min/max with dim reduction is nonlinear.  The returned index tensor is the culprit.

## solution
Created custom handler to compute local reductions with global indices.  Use the min/max values to perform a linear reduction locally.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @aditew01 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @Lucaskabela @xmfan

### CODE REVIEW

**Comment by skpark-rh** [comment]
Yeah @zpcore reviewed my last one with argmin/argmax so he'll me more familiar with this code. FYI I refactored the previous implementation to make it a little bit more reusable!

File: torch/distributed/tensor/_nonlinear_redux.py [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+        self._keepdim: bool = False
+        self._shard_mesh_dims: list[int] = []
+
+    def __call__(
+        self,
+        op_call: torch._ops.OpOverload,
+        args: tuple[object, ...],
+        kwargs: dict[str, object],
+    ):
+        raise NotImplementedError
```
Comment (zpcore) [inline torch/distributed/tensor/_nonlinear_redux.py:34]: What about defining as an abc class and wrapping with `abstractmethod`?

File: torch/distributed/tensor/_nonlinear_redux.py [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+
+    def _gather_tensors(
+        self,
+        gather_dim: int,
+        gathered_idxs: torch.Tensor,
+        local_redux: torch.Tensor,
+        device_mesh: torch.distributed.device_mesh.DeviceMesh,
+    ) -> tuple[torch.Tensor, torch.Tensor]:
```
Comment (zpcore) [inline torch/distributed/tensor/_nonlinear_redux.py:48]: Comment that this is gathering both (min or max) tensor value and the corresponding index.

File: torch/distributed/tensor/_nonlinear_redux.py (hunk 1 above)
Comment (skpark-rh) [reply torch/distributed/tensor/_nonlinear_redux.py:34]: Yeah I thought about that actually. I'll get that squared away.

File: torch/distributed/tensor/_nonlinear_redux.py (hunk 2 above)
Comment (skpark-rh) [reply torch/distributed/tensor/_nonlinear_redux.py:48]: Okay I'll add that comment in the method.

File: torch/distributed/tensor/_nonlinear_redux.py [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+    ) -> None:
+        for mesh_dim, p in enumerate(placements):
+            if isinstance(p, Shard):
+                if self._dim is None or p.dim == (
+                    self._dim if self._dim >= 0 else local_tensor.ndim + self._dim
+                ):
+                    self._shard_mesh_dims.append(mesh_dim)
+        return
+
+    @staticmethod
```
Comment (zpcore) [inline torch/distributed/tensor/_nonlinear_redux.py:175]: I think we don't need this staticmethod. Also, we can call this single line `unwrap_to_op_info` directly.

Comment (skpark-rh) [reply torch/distributed/tensor/_nonlinear_redux.py:175]: I didn't want to duplicated code in the child classes below but if writing this out explicitly in the `__call__` method helps communicate things I can make this change.

File: torch/distributed/tensor/_nonlinear_redux.py [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+                gathered_idxs,
+                gather_dim=gather_dim,
+                group=(device_mesh, mesh_dim),
+            )
+        return gathered_redux, gathered_idxs
+
+    def _convert_to_global_idxs(
+        self,
+        local_idx: torch.Tensor,
+        local_tensor: torch.Tensor,
```
Comment (zpcore) [inline torch/distributed/tensor/_nonlinear_redux.py]: Do we need `local_tensor` here? `local_tensor` shape can be returned from compute_local_shape_and_global_offset.

Comment (skpark-rh) [reply torch/distributed/tensor/_nonlinear_redux.py]: You are absolutely right! Good catch!

Comment (skpark-rh) [reply torch/distributed/tensor/_nonlinear_redux.py]: I do need the device though which I can pass as a parameter.

File: torch/distributed/tensor/_nonlinear_redux.py [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+            local_coord = torch.unravel_index(local_idx, local_tensor.shape)
+            global_coord = torch.stack(local_coord)
+            gather_dim = 0
+            for i, offset in enumerate(global_offset):
+                global_coord[i] += offset
+            # compute with proper striding
+            gathered_idxs = torch.tensor(
+                0, device=local_tensor.device, dtype=torch.long
+            )
+            for i, coord in enumerate(global_coord):
```
Comment (zpcore) [inline torch/distributed/tensor/_nonlinear_redux.py:97]: We can do this way:
```
accum_size = 1
for i in reversed(range(len(global_offset))):
    gathered_idxs += coord * accum_size
    accum_size *= global_shape[i]
```

Comment (skpark-rh) [reply torch/distributed/tensor/_nonlinear_redux.py:97]: I need the slice of the global_shape. Only grabbing the ith element will not compute the stride correctly.

File: torch/distributed/tensor/_nonlinear_redux.py (hunk 4 above)
Comment (zpcore) [reply torch/distributed/tensor/_nonlinear_redux.py]: Can we get the device from `local_idx`?

File: torch/distributed/tensor/_nonlinear_redux.py (hunk 5 above)
Comment (skpark-rh) [reply torch/distributed/tensor/_nonlinear_redux.py:97]: Oh actually since you are reversing it, this is the same thing.

Comment (skpark-rh) [reply torch/distributed/tensor/_nonlinear_redux.py:97]: The `coord` will have to be indexed from the `global_coord` though.

File: torch/distributed/tensor/_nonlinear_redux.py (hunk 4 above)
Comment (skpark-rh) [reply torch/distributed/tensor/_nonlinear_redux.py]: Yep was just about to comment that. lol

File: torch/distributed/tensor/_nonlinear_redux.py [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+
+    def _collect_shard_mesh_dims(
+        self, local_tensor: torch.Tensor, placements: tuple[Placement, ...]
+    ) -> None:
+        for mesh_dim, p in enumerate(placements):
+            if isinstance(p, Shard):
+                if self._dim is None or p.dim == (
+                    self._dim if self._dim >= 0 else local_tensor.ndim + self._dim
+                ):
+                    self._shard_mesh_dims.append(mesh_dim)
```
Comment (zpcore) [inline torch/distributed/tensor/_nonlinear_redux.py:168]: Add the elif condition to error out. We will isolate `Shard` and `_StridedShard` in upcoming PRs. I don't know how `_StridedShard` will impact the index result...
```
if isinstance(p, Shard):
    ...
elif isinstance(p, _StridedShard):
    raise f"{self.op_call}" doesn't support _StridedShard.
```

Comment (skpark-rh) [reply torch/distributed/tensor/_nonlinear_redux.py:168]: Ah I see! I will add that! Thanks for the heads up!

File: torch/distributed/tensor/_nonlinear_redux.py [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+        placements: tuple[Placement, ...],
+    ) -> tuple[int, torch.Tensor]:
+        _, global_offset = compute_local_shape_and_global_offset(
+            global_shape, device_mesh, placements
+        )
+
+        if self._dim is None:
+            local_coord = torch.unravel_index(local_idx, local_tensor.shape)
+            global_coord = torch.stack(local_coord)
+            gather_dim = 0
```
Comment (zpcore) [inline torch/distributed/tensor/_nonlinear_redux.py:92]: Why do we return `gather_dim=0` here?

File: torch/distributed/tensor/_nonlinear_redux.py (hunk 5 above)
Comment (zpcore) [reply torch/distributed/tensor/_nonlinear_redux.py:97]: This should also work. I think your code is also fine though since `global_shape` is not large.
```
accum_size = 1
for coord, dim_size in reversed(zip(global_coord, global_shape)):
    gathered_idxs += coord * accum_size
    accum_size *= dim_size
```

File: torch/distributed/tensor/_nonlinear_redux.py (hunk 7 above)
Comment (skpark-rh) [reply torch/distributed/tensor/_nonlinear_redux.py:92]: For `dim = None`, the tensor is flattened and a single value/index is returned. I put the `gather_dim=0` to collect all the operated values from different processes to the 0th dimension. I call the `op_call` again in `__call__` of the child classes to get the rank winners on the 0th dim to get the true winning number.
